internal/consensus/state.go

```diff
@@ -1346,6 +1369,32 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	if cs.Proposal == nil {
```

Comment:
Any way we can unit-test this?

Reply:
Also (now that I remember). I this condition is necessary in order for line 1378 not to panic. In other words, the logic implemented in line 1378 assumes we have a proposal and a proposed block. And that logic seems to me it's core PBTS logic

Reply:
Yes, I see that, although L1378 could be changed as well.  I also realized that in line 1618 (precommit) we do the same.

The point here is that a block backed by 2/3+ prevotes is good. So we would not need to perform all checks for it. And the `timely` check is performed on the proposal.

 This is a change in the implementation of the protocol. It in general assumes that receiving a `Block` is enough for prevoting/precommiting it. This because the propagation of Proposal and Block are independent. This discussion however is more complex, should I open an issue for that?

Reply:
Please do ahead and open an issue. Once you're done, please add a reference to that issue in the list of "Additional Changes/Revisions" in #1731

Reply:
#2119